commit 060c7109cfee0f43db5b05d3bcdae867539ece38
Author: Chris Feist <cfeist@redhat.com>
Date:   Mon Aug 2 16:59:07 2010 -0500

    Removed extra logging from luci speedup code.

diff --git a/luci/controllers/cluster.py b/luci/controllers/cluster.py
index a3a90ec..62adfa2 100644
--- a/luci/controllers/cluster.py
+++ b/luci/controllers/cluster.py
@@ -83,7 +83,6 @@ class IndividualClusterController(BaseController):
         identity = request.environ.get('repoze.who.identity')
         self.username = identity['repoze.who.userid']
 
-        log.debug("Creating %s" % repr(self))
         tmpl_context.show_sidebar = True
         tmpl_context.cluster_name = self.name
         tmpl_context.cluster_url = '/cluster/%s/' % self.name
@@ -105,7 +104,6 @@ class IndividualClusterController(BaseController):
 
     def get_model(self):
         if not self.model:
-            log.debug("Obtaining cluster model for %s" % repr(self))
             self.model = get_model_for_cluster(self.name, self.get_agent())
         return self.model
 
